### PR TITLE
EventHubStreamProvider flow control

### PR DESCRIPTION
**Pull request by jason-bragg**
Introduced basic flow control mechanism to Persistent Stream Provider's adapter.
Added minimal flow control to EventHubStreamProvider that uses relative positions of cursors in cache to detect risk of cache miss errors, and reduces read rate in those cases.

Possible flow control mechanisms under consideration:
Load Shedding - Use silo load shedding detection to turn off queue readers.
Event/Second upper bound - Use configurable max event per second read rate.

This change minimally addresses Flow Control described in Recoverable Event Hub Persistent Stream Provider - Part 2 #1454

### CODE REVIEW

File: src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs [hunk 1]
```diff
@@ -8,6 +8,13 @@ public interface IEventHubSettings
         string ConsumerGroup { get; }
         string Path { get; }
         int? PrefetchCount { get; }
+
+        /// <summary>
+        /// Indicates if stream provider should read all new data in partition, or from start of partion.
```
Comment (jason-bragg) [inline src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs]: spelling: partion

Comment (veikkoeeva) [inline src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs]: http://www.thefreedictionary.com/partition
http://www.urbandictionary.com/define.php?term=partion

File: src/Orleans/Streams/QueueAdapters/AggregatedQueueFlowControler.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Orleans.Streams
+{
+    public class AggregatedQueueFlowControler : List<IQueueFlowControler>, IQueueFlowControler
```
Comment (gabikliot) [inline src/Orleans/Streams/QueueAdapters/AggregatedQueueFlowControler.cs]: controller with double l

**Comment by gabikliot** [comment]
How that differs from the notion of backpressure that we have in Azure Queue adapter?

**Comment by jason-bragg** [comment]
@gabikliot 

>How that differs from the notion of backpressure that we have in Azure Queue adapter? 

Same thing, just rebranded and generalized.  While reexamining the backpressure regulator in the 343 EventHubStreamProvider, prior to porting it over, the general consensus was that 'flow control' was a clearer term for this behavior.  It was also concluded that we should start with an implementation that is simpler than the one in 343, and only add more complexity if the simpler approach acts poorly in testing.

**Comment by gabikliot** [comment]
@jason-bragg , can you please explain with a bit more the "same thing".
Solves eventually the same problem, sure, but a different mechanism I presume. So can you please describe how it works, how it differs, why better, is there a plan to unify both?

**Comment by jason-bragg** [comment]
@gabikliot 
>Solves eventually the same problem

Correct.  That's what was intended by "Same thing, just rebranded and generalized".  Backpressure controlled the rate in which events were ingested from the queue, which is probably better understood as 'flow control'.  Since we've not yet (and may not be able to) settled on a single pattern of flow control that works for all queues and stream processing needs, we introduce flow control as a general feature of the adapter framework, primarily exposed through the cache, as that was where the capability was formerly exposed.

>So can you please describe how it works, how it differs.

In addition to generalizing the 'flow control' behavior, a different flow control mechanism was introduced for the EventHubStreamProvider.  This flow control is intended for real time systems.  It is intended to reduce the read rate based on how far behind consumers become.  The main difference in this behavior is that if a single (or relatively small set) of consumers get behind, they will not slow down the vast majority of consumers that are keeping up with the read rate.  This is appropriate for a real time system where there is a basic assumption that the older data is, the less important it is, hence losing old data is preferred to delaying processing of new data.  This is very different from the flow control mechanism in the SimpleQueueCache which will slow the read rate to match the slowest consumer.

More specifically, each time an event for a consumer is found in the cache a measurement is made gauging how far behind the most recent data that event is.  This measurement contributes to the overall 'pressure' the cache is under.  If this measurement is beyond a specific threshold, the agent will stop reading from the queue until the consumers read enough to reduce the pressure.  For instance:

100 events in the cache (positions 1-100, where 100 is the most recent data), with pressure threshold of 0.3
Consumer 1 consumes two events one at positions 95, then 92.
This would contribute to pressure :
    95 => 100-95/100 = 0.05
    92 => 100-92/100 = 0.08
Consumer 2 consumes one event at position 42.
This would contribute to pressure :
    42 => 100-42/100 = 0.58

Overall pressure would be:  (0.05 + 0.08 + 0.58) / 3 = ~0.236
Even though consumer 2 is quite far pass the threshold, it would not cause the read rate to slow as the overall pressure is below 0.3.

However, if Consumer 2 was more behind:
Consumer 2 consumes one event at position 21.
This would contribute to pressure :
    21 => 100-21/100 = 0.79

Overall pressure would be:  (0.05 + 0.08 + 0.79) / 3 = ~0.306
In this case reading would stop until the consumers got caught up.

However, in the case of many consumers which are keeping up, it is possible for the rate to not slow down no matter how far back a single consumer is:
Consider the following contributions to back pressure.  .02, .04, .04, .05, .09, .12, .15, .99.
Even though the consumer contributing measurement 0.99 is about to lose data because they are at the end of the cache, most consumers are fine (total pressure of ~0.187), so reading will continue as normal.

>why better

The flow control mechanism used in the EventHubStreamProvider is not 'better' than the mechanism in the SimpleQueueCache, it is simply different.  It serves the same purpose, flow control, but for a different set of requirements.

>is there a plan to unify both?

I have no such plan, though it would be nice to do so.  It is my hope that as more applications use the streaming infrastructure, we'll see patterns form which will inform us on what flow control mechanisms are most common.  Once these common mechanisms are identified, we can add them to Orleans Streaming.  If these mechanisms converge to a single pattern that works for everyone, then we can settle on that pattern, and simplify around that single solution.  The idea here is to allow customization until we learn what behaviors are best rather than trying to forecast with relatively little information.

**Comment by gabikliot** [comment]
Thanks for the details explanation @jason-bragg !

:+1:

File: src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs [hunk 3]
```diff
@@ -34,6 +34,26 @@ public class PooledQueueCache<TQueueMessage, TCachedMessage>
         private readonly ICacheDataAdapter<TQueueMessage, TCachedMessage> cacheDataAdapter;
         private readonly ICacheDataComparer<TCachedMessage> comparer;
 
+        public TCachedMessage? Newest
+        {
+            get
+            {
+                if (IsEmpty)
+                    return null;
+                return messageBlocks.First.Value.NewestMessage;
```
Comment (jdom) [inline src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs]: this is probably not multi-threaded, but just in case (and also reads easier) now that we use c# 6 you can write:
`return messageBlocks.First?.Value.NewestMessage;`
without the need to check for empty or null

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs [hunk 4]
```diff
@@ -2,35 +2,96 @@
...
+            cacheDataAdapter.PurgeAction = cache.Purge;
+            cache.OnPurged = CheckpointOnPurged;
+
+            cachePressureMonitor = new CachePressureMonitor();
         }
 
+        protected abstract string GetOffset(TCachedMessage lastItemPurged);
+
+        // cachePressureContribution should be a double between 0-1, indicating how much danger the item is of being removed from the cache.  0 indicating  no danger, 1 indecating removal is imminent.
+        protected abstract bool TryCalculateCachePressureContribution(StreamSequenceToken token, out double cachePressureContribution);
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: Convert this to an XML comment. Also typo in "indecating"

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs [hunk 5]
```diff
@@ -2,35 +2,96 @@
...
+            }
+            return true;
+        }
+
+        private class CachePressureMonitor
+        {
+            const double pressureThreshold = 1.0/3.0;
+
+            private double accumulatedCachePressure;
+            private int cachePressureContributionCount;
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: This seems to rigid. Are there plans to make this more configurable? Also we talked about having different aggregation functions (such as max or min instead of just average, not saying they need to be implemented now) so maybe the class name should reflect the method used?

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs [hunk 6]
```diff
@@ -2,35 +2,96 @@
...
+        }
+
+        public bool TryGetNextMessage(object cursorObj, out IBatchContainer message)
+        {
+            if (!cache.TryGetNextMessage(cursorObj, out message))
+                return false;
+            double cachePressureContribution;
+            if (TryCalculateCachePressureContribution(message.SequenceToken, out cachePressureContribution))
+            {
+                cachePressureMonitor.RecordCachePressureContribution(cachePressureContribution);
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: if I understand correctly, you'll be recording values every time you read, and only once in a while averaging and resetting the recorded values. That could mean that a fast reader will record very often, and slow readers might not even record once before the values reset. Is that correct? Then the math is not exactly as explained, where each reader's weigth is the same as others. That is not necessarily wrong, as it might just mean that we need to lower/tune the threshold to account for it, but I'm pointing it out in case there is something a little bit better we could do, such as recording the value associated to each reader once (always overwriting the previous if there's any).

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs [hunk 7]
```diff
@@ -2,35 +2,96 @@
...
         }
 
-        protected abstract string GetOffset(TCachedMessage lastItemPurged);
-
         private void CheckpointOnPurged(TCachedMessage lastItemPurged)
         {
             Checkpointer.Update(GetOffset(lastItemPurged), DateTime.UtcNow);
         }
+
+        public int MaxAddCount { get { return cachePressureMonitor.IsUnderPressure() ? 0 : defaultMaxAddCount; } }
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: this property is not side-effects free, please convert to a method.

File: src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs [hunk 8]
```diff
@@ -8,6 +8,13 @@ public interface IEventHubSettings
...
         string Path { get; }
         int? PrefetchCount { get; }
+
+        /// <summary>
+        /// Indicates if stream provider should read all new data in partition, or from start of partition.
+        /// True - read all new data added to partition.
+        /// False - start reading from beginning of partition.
+        /// Note: If checkpoints are used, stream provider will always begin reading from most recent checkpoint.
+        /// </summary>
         bool StartFromNow { get; }
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs]: thanks for the comments. Nevertheless, this highlights whether it's appropriate to use a bool flag instead of a more explicit enum. No need address now though as it's not a big deal, just something I forgot to comment on when this was initially added, and this addition reminded me.

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs (hunk 7 above)
Comment (jason-bragg) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: the MaxAddCount is part of the stream adapter's public api.  I was trying not to introduce breaking changes, but you're correct, this probably shouldn't be a property.

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs (hunk 6 above)
Comment (jason-bragg) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: Your understanding of this is correct.  In the more detailed description I provided, at Gabriel's request, I included two streams, one which read faster and contributed more to the pressure.  I included this example to help illuminate such peculiarities.

This approach was selected because it's fast and simple.  Testing is needed to determine if it's a viable algorithm.

Most alternative approaches I've explored come with a cost.  For instance, tracking each cursor's most recent contribution incurs the cost of a heavier data structure (dictionary), a lookup per contribution, a more expensive aggregation, and a more expensive reset after each aggregated measurement.  While these costs probably wouldn't effect performance noticeably, they are overhead that shouldn't be incurred unless we've reason to believe lighter algorithms are not sufficient.

I'm very open to alternatives if there is reason to believe the value of the measurement relative to the cost of the operation is better.  In my view this is a very interesting space to explore, but given my current timelines, I will likely settle for the first fast simple algorithm that works reasonably well.

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs (hunk 5 above)
Comment (jason-bragg) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: The original algorithm was more complex, and tunable, and virtually incomprehensible to most users. :(

I'm intentionally erring on the side of simplicity this time.  Hopefully load testing will determine if this algorithm is viable and if configurable tuning knobs are necessary.

As for supporting different algorithms, the initial structure was setup to support that as a natural extension, but flow control has not yet been exposed as an extension point of the EventHubStreamProvider, as data format, checkpointing, and delivery failure handlers are.  The reason this has not been made an extension point is the hopes that a simple default implementation will be suffice.  Each extension point in the provider adds to the complexity, so I’m trying hard not to introduce extension points unless they are absolutely necessary.

File: src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs (hunk 8 above)
Comment (jason-bragg) [inline src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs]: Agreed.  I've played with the idea of using enums or enum flags, but not yet convinced myself that the alternatives are better.

Open to suggestions if there are alternatives that are arguably better.  A change to this would be better as a separate PR though, so if you have suggestions, please post and if better approach is settled on, I’ll submit in separate PR.

File: src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs (hunk 3 above)
Comment (jason-bragg) [inline src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs]: >this is probably not multi-threaded

Yes, cache expects single threaded access.

>without the need to check for empty

Unfortunately the IsEmpty is a bit more involved than just a null check.

Good to know we can use c6 now though.

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs (hunk 5 above)
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: you know where I stand, I don't believe a default that can drop messages for the slow consumers is something that will fit 100% of the scenarios.
A lot of people will prefer to just be reading messages a few seconds/minutes older when there is a spike in commands being processed until it catches back up. Not everyone needs sub-second turnaround the entire time. I can see the dropped messages being an issue for people doing Event Sourcing.

Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: But again, not saying you must add the additional capabilities in this PR, just saying that we should acknowledge the fact that there won't be a single default that everyone likes, so we might as well name this implementation appropriately and not as if it will be the single one that needs no qualification.

**Comment by jdom** [comment]
Other than the few remaining comments, this looks good

File: src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs (hunk 8 above)
Comment (veikkoeeva) [inline src/OrleansServiceBus/Providers/Streams/EventHub/IEventHubSettings.cs]: If it matters, for an "untrained eye" it would look more clear with an enum. Maybe something like `ReadFromBeginning`, `ReadFromCheckpoint` (or what could be the word what's new data only).

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs (hunk 5 above)
Comment (jason-bragg) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: > I don't believe a default that can drop messages for the slow consumers is something that will fit 100% of the scenarios.

I agree.  We'll likely need to make an extension point for that at some point, but this behavior 'should' suite all current customers, and adding an extension point would be a natural extension that could be made when needed.

To community looking to use the event hub stream provider:  Now would be a good time to chime in on this.  Is this worth opening an issue to discuss?

> might as well name this implementation appropriately

Fair point.  The generic name makes it seem like it's the only expected implementation.   Should probably also move it out of being a nested class.  Ideas for name?  AveragingCachePressureMonitor?
